compiler/lib/ir.ml

```diff
@@ -0,0 +1,158 @@
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, with linking exception;
+ * either version 2.1 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ *)
+open Stdlib
+
+type array_literal = element_list
+
+and element_list = expression option list
+
+and binop =
+  | Eq
+  | StarEq
+  | SlashEq
+  | ModEq
+  | PlusEq
+  | MinusEq
+  | Or
+  | And
+  | Bor
+  | Bxor
+  | Band
+  | EqEq
+  | NotEq
+  | EqEqEq
+  | NotEqEq
+  | InstanceOf
+  | Lsl
+  | Lsr
+  | Asr
+  | Plus
+  | Minus
+  | Mul
+  | Div
+  | Mod
+  | Lt
+  | Le
+  | Gt
+  | Ge
+  | FloatPlus
+  | FloatMinus
+  | FloatMul
+  | FloatDiv
+  | FloatMod
+  | FloatLt
+  | FloatLe
+  | FloatGt
+  | FloatGe
+  | FloatEqEq
+  | FloatNotEq
+
+and unop =
+  | Not
+  | Neg
+  | FloatNeg
+  | Typeof
+  | IsInt
+  | ToInt
+  | ToBool
+  | IntToString
+  | FloatToInt
+  | IntToFloat
+  | Void
+  | Delete
+  | Bnot
+
+and arguments = expression list
+
+and property_name_and_value_list = (Id.identifier * expression) list
+
+and expression =
+  | ERaw of string
+  | ESeq of expression * expression
+  | ECond of expression * expression * expression
+  | EBin of binop * expression * expression
+  | EUn of unop * expression
+  | ECall of expression * arguments * Loc.t
+  | ECopy of expression * Loc.t
+  | EVar of Id.t
+  | EFun of function_expression
+  | EArityTest of expression
+  | EStr of string * [ `Bytes | `Utf8 ]
+  | EVectlength of expression
+  | EArrAccess of expression * expression
+  | EArrLen of expression
+  | EArr of array_literal
+  | EStructAccess of expression * int
+  | EStruct of arguments
+  | ETag of int * arguments
+  | EDot of expression * Id.identifier
+  | EAccess of expression * expression
+  | ENew of expression * arguments option
+  | EObj of property_name_and_value_list
+  | EBool of bool
+  | EFloat of float
+  | EInt of int
+  | EQuote of string
+  | ERegexp of string * string option
+  | ERuntime
+
+and statement =
+  | Block of block
+  | Variable_statement of variable_declaration list
+  | Empty_statement
+  | Expression_statement of expression
+  | If_statement of expression * (statement * Loc.t) * (statement * Loc.t) option
+  | Loop_statement of statement * Loc.t
+  | Continue_statement of Javascript.Label.t option
+  | Break_statement of Javascript.Label.t option
+  | Return_statement of expression option
+  | Labelled_statement of Javascript.Label.t * (statement * Loc.t)
```

Comment:
Should the IR contain primitives like this that aren't present in OCaml?

Reply:
These control flow structures are well recognized and exist in some form in all high-level languages. What would you suggest for the IR? I didn't want to make non-cosmetic changes to `generate.ml`, at least in this PR. These primities don't exist in OCaml, but control flow deconstruction produces them right now and they are not too much specific for JavaScript (in Golang backend, these got transformed to Go AST seamlessly).